src/main/java/spoon/support/util/internal/lexer/JavaLexer.java

```diff
@@ -0,0 +1,399 @@
+			}
+		};
+	}
+
+	private Token createToken(TokenType type, int pos) {
+		return new Token(type, this.charRemapper.remapPosition(pos), this.charRemapper.remapPosition(this.nextPos));
+	}
+
+	private Token angleBracket(int pos, int found, int maxFound, char bracket) {
+		if (hasMore()) {
+			char peek = peek();
+			if (peek == '=') {
+				next();
+				return createToken(TokenType.OPERATOR, pos);
+			}
+			if (peek == bracket && found < maxFound) {
+				next();
+				return angleBracket(pos, found + 1, maxFound, bracket);
+			}
+		}
+		return createToken(TokenType.OPERATOR, pos);
+	}
+
+	private void skipUntilLineBreak() {
+		while (hasMore()) {
+			char peek = peek();
+			if (peek == '\n' || peek == '\r') {
+				skipWhitespaces();
+				return;
+			}
+			next();
+		}
+	}
+
+	private boolean skipUntil(String s) {
+		char[] chars = s.toCharArray();
+		char first = chars[0];
+		int pos = this.nextPos;
+		do {
+			int index = indexOf(first, pos);
+			if (index < 0) {
+				// TODO what if not present?
+				return false;
+			}
+			if (Arrays.equals(this.content, index, index + chars.length, chars, 0, chars.length)) {
+				this.nextPos = index + chars.length;
+				return true;
+			} else {
+				pos = index + 1;
+			}
+		} while (true);
+	}
+
+	private @Nullable Token lexLiteralOrKeywordOrIdentifier() {
+		int pos = this.nextPos;
+		char next = next(); // assuming next is available
+		if (Character.isJavaIdentifierStart(next)) {
+			while (hasMore() && isJavaIdentifierPart(peek())) {
+				next();
+			}
+			Optional<JavaKeyword> match = KEYWORD_TRIE.findMatch(this.content, pos, this.nextPos);
+			if (match.isPresent()) {
+				return createToken(TokenType.KEYWORD, pos);
+			}
+			// special case: non-sealed is not a valid java identifier, but a keyword
+			if (isNonSealed(pos)
+			) {
+				skip("sealed".length() + 1); // move behind
+				return createToken(TokenType.KEYWORD, pos);
+			}
+			return createToken(TokenType.IDENTIFIER, pos);
+		} else if (Character.isDigit(next)) {
+			readNumericLiteral(next);
+			return createToken(TokenType.LITERAL, pos);
+		}
+		return null;
+	}
+
+	private boolean isNonSealed(int pos) {
+		int requiredForNonSealed = "sealed".length();
+		boolean startsWithNonSealed = hasMore(requiredForNonSealed)
+				&& isNon(pos, this.content)
+				&& isDashSealed(content, this.nextPos);
+		if (startsWithNonSealed) {
+			return !hasMore(requiredForNonSealed + 1)
+					|| !isJavaIdentifierPart(peek(requiredForNonSealed + 1));
+		}
+		return false;
+	}
+
+	private void readNumericLiteral(char first) {
+		if (!hasMore()) {
+			return;
+		}
+		// check if hexadecimal notation
+		if (first == '0') {
+			if (peek() == 'x' || peek() == 'X') {
+				next();
+				readHexadecimalsAndUnderscore();
+				if (hasMore() && peek() == '.') {
+					next();
+					readHexadecimalFloatingPointLiteral();
+				} else {
+					readHexadecimalsAndUnderscore();
+				}
+			} else if (peek() == 'b' || peek() == 'B') {
+				next();
+				readDigitsOrUnderscore();
```

Comment:
Do we want to restrict this to binary? I guess constraining it would make the lexer less robust and range checking for binary and octal isn't your job?